### PR TITLE
Add CBOR diagnostic files for ease of code review

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes https://github.com/dhall-lang/dhall-lang/issues/656

This adds a human-readable `.diag` file containing the CBOR diagnostic
representation for the corresponding `.dhallb` file, for ease of
reviewing changes to encoded expressions.

This also adds:

* A new `./scripts/generate-diagnostic-files.sh` script for keeping the
  `*.diag` files up-to-date
* A new CI check to ensure the contributor hasn't forgotten to update those
  files

This also removes the `./tests/diff-binary.sh` script since it's no longer
necessary.

### CODE REVIEW

**Comment by sjakobi** [comment]
CI seems to work:

```
Only in ./tests.expected/parser/success/unit: ListLitEmptyPrecedenceB.diag
diff --recursive ./tests.actual/parser/success/unit/import/urls/potPourriB.diag ./tests.expected/parser/success/unit/import/urls/potPourriB.diag
1c1
< [24, null, 0, 1, null, "-._~%2C!$&'*+;=:@0abc1--12a------a-a--a-0", "foo", "/-._~%2C!$&%27*+;=:@/?"]
---
> [24, null, 0, 1, null, "-._~%2C!$&'*+;=:@0abc1--12a------a-a--a-0", "foo", "/-._~%2C!$&'*+;=:@/?"]
builder for '/nix/store/fckqqq62vi62yw8v1f1g36j8fzl2yy3d-diagnostic-files-lint.drv' failed with exit code 1
```

:+1:

File: tests/parser/success/unit/import/urls/potPourriB.diag [hunk 1]
```diff
@@ -0,0 +1 @@
+[24, null, 0, 1, null, "-._~%2C!$&'*+;=:@0abc1--12a------a-a--a-0", "foo", "/-._~%2C!$&%27*+;=:@/?"]
```
Comment (sjakobi) [inline tests/parser/success/unit/import/urls/potPourriB.diag]:
```suggestion
[24, null, 0, 1, null, "-._~%2C!$&'*+;=:@0abc1--12a------a-a--a-0", "foo", "/-._~%2C!$&'*+;=:@/?"]
```

File: release.nix [hunk 2]
```diff
@@ -73,6 +84,15 @@ let
       done
     '';
 
+    diagnostic-files-lint = pkgsNew.runCommand "diagnostic-files-lint" {} ''
+      ${pkgsNew.rsync}/bin/rsync --archive ${pkgsNew.expected-diagnostic-files}/ ./tests.expected
+      ${pkgsNew.rsync}/bin/rsync --archive ${./tests}/ ./tests.actual
+
+      ${pkgsNew.diffutils}/bin/diff --recursive ./tests.{actual,expected}
```
Comment (ocharles) [inline release.nix]: Don't you need to do something here to assert that there are no differences? Or does `diff` exit with non-zero if there are differences?

Comment (Gabriel439) [reply release.nix]: @ocharles: `diff` exits with a non-zero exit code if there are differences

**Pull request merged**